File: python/endstone_sign/events.py
from __future__ import annotations

from collections.abc import Callable
from threading import RLock

from .model import SignEvent


SignEventListener = Callable[[SignEvent], None]


class SignEventBus:
    def __init__(self) -> None:
        self._lock = RLock()
        self._listeners: dict[int, SignEventListener] = {}
        self._next_id = 1

    def add_listener(self, listener: SignEventListener) -> int:
        if not callable(listener):
            raise TypeError("sign event listener must be callable")
        with self._lock:
            listener_id = self._next_id
            self._next_id += 1
            self._listeners[listener_id] = listener
            return listener_id

    def remove_listener(self, listener_id: int) -> bool:
        with self._lock:
            return self._listeners.pop(listener_id, None) is not None

    def publish(self, event: SignEvent) -> None:
        with self._lock:
            listeners = tuple(self._listeners.values())
        for listener in listeners:
            listener(event)
            if event.cancellable and event.cancelled:
                break

File: src/events.cpp
#include "endstone_sign/events.h"

#include <algorithm>
#include <stdexcept>
#include <utility>

namespace endstone_sign {

std::size_t SignEventBus::addListener(SignEventListener listener) {
    if (!listener) throw std::invalid_argument("sign event listener must not be empty");
    std::scoped_lock lock(mutex_);
    const auto id = next_id_++;
    listeners_.push_back({id, std::move(listener)});
    return id;
}

bool SignEventBus::removeListener(std::size_t id) {
    std::scoped_lock lock(mutex_);
    const auto old_size = listeners_.size();
    std::erase_if(listeners_, [id](const ListenerEntry &entry) { return entry.id == id; });
    return listeners_.size() != old_size;
}

void SignEventBus::publish(SignEvent &event) const {
    std::vector<ListenerEntry> listeners;
    {
        std::scoped_lock lock(mutex_);
        listeners = listeners_;
    }
    for (const auto &entry : listeners) {
        entry.listener(event);
        if (event.cancellable && event.cancelled) break;
    }
}

} // namespace endstone_sign
